accepted/2023/nuget-config-apikeys-unencrypted.md

```diff
@@ -0,0 +1,115 @@
+- GitHub PR (not yet implemented)
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+
+Enable unencrypted API keys in `nuget.config` files, so that they can be used on platforms other than Windows.
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+
+Currently `nuget.config` `<apikeys>` section only supports encrypted values, via .NET's [`System.Security.Cryptography.ProtectedData` APIs](https://learn.microsoft.com/dotnet/api/system.security.cryptography.protecteddata).
+As the important message says on the docs page, on Windows it uses Windows' Data Protection API (DPAPI), which isn't supported on Linux or Mac, and [the .NET libraries team have no plans to support it since it will have different security promises than on Windows](https://github.com/dotnet/runtime/issues/22886).
+There is [an issue to consider another way of having less secure "encrypted" passwords](https://github.com/NuGet/Home/issues/1851) on Linux and Mac (like what Mono does for the `ProtectedData` API), but it hasn't been implemented yet.
+Therefore, in order for this feature not to be blocked, `<apikeys>` needs to support unencrypted (plain text) secrets, just as source credentials do.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+Add an XML attribute `encrypted`, which will be a boolean `true`/`false`, to specify if the value is encrypted or not.
+If omitted from the element, the default value is true, to maintain backwards compatibility with prior versions of NuGet.
+For example:
+
+```xml
+<apikeys>
+    <add key="https://api.nuget.org/v3/index.json" value="api_key" encrypted="false" />
+</apikeys>
+```
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+We have got [feedback from customers](https://github.com/NuGet/Home/issues/3245) who manually entered clear text package credentials where NuGet expects encrypted credentials, and NuGet will crash with an obscure error message ("The parameter is incorrect") that is not actionable by customers.
+While we are working on API keys, I propose that we should additionally catch the decryption of encrypted API keys, and if DPAPI throws an exception, report a meaningful error message back to the customer.
+
+The `<packageSources>` section also uses the `<add key="string" value="string">` syntax that most other parts of `nuget.config` use, but it also supports the `protocolVersion="int"` attribute.
+The approach it takes to add `protocolVersion` can be copied to make the `encrypted` attribute work.
+
+## Drawbacks
+
+<!-- Why should we not do this? -->
```

Comment:
You know why :) So fill this out.

- Users run the risk of checking the file into source control.
- Users run the risk of putting this into a directory not properly ACLed against other users on the machine
- Users run the risk of malware on their machine reading the credentials and sending them elsewhere.

Reply:
I would assume that the last point is not mitigated by having encrypted API keys (if nuget client can decrypt it, so can malware), so it should be irrelevant in this context.

Reply:
The "Drawbacks" section is a 2nd level header, and I added two 3rd level headers under. If you look at the rendered spec, it's much more clear. I believe these points are addressed in the "unencrypted secrets" sub-heading, but I did update the spec last week which hopefully made it a little better as well.